packages/devtools_app/lib/src/screens/memory/framework/offline_data/offline_data.dart

```diff
@@ -0,0 +1,59 @@
+// Copyright 2024 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '../../panes/chart/controller/chart_pane_controller.dart';
+import '../../panes/diff/controller/diff_pane_controller.dart';
+import '../../panes/profile/profile_pane_controller.dart';
+import '../../shared/heap/class_filter.dart';
+
+class _Json {
+  static const selectedTab = 'selectedTab';
+  static const classFilter = 'classFilter';
+  static const diffData = 'diffData';
+  static const profileData = 'profileData';
+  static const chartData = 'chartData';
+}
+
+class OfflineMemoryData {
+  OfflineMemoryData(
+    this.diff,
+    this.profile,
+    this.chart,
+    this.filter, {
+    this.isEmpty = false,
+    required this.selectedTab,
+  });
+
+  factory OfflineMemoryData.parse(Map<String, dynamic> json) {
+    Map<String, dynamic> item(String key) =>
+        json[key] as Map<String, dynamic>? ?? {};
+    return OfflineMemoryData(
+      DiffPaneController.parse(item(_Json.diffData)),
+      ProfilePaneController.parse(item(_Json.profileData)),
+      MemoryChartPaneController.parse(item(_Json.chartData)),
+      ClassFilter.parse(item(_Json.classFilter)),
+      selectedTab: json[_Json.selectedTab] as int? ?? 0,
+      isEmpty: json.isEmpty,
```

Comment:
It is easier to process data than to detect if it is empty. To many defaults to verify they are equal to the values. So, for memory screen I will always return false.

Reply:
Fine to return false for now, but can you add a TODO to implement this in the future? Once all the serialization work is done, we should have a simple way to do this.

Reply:
Why it is important to have empty data as a special case? If data is empty, processing of data will not save any performance, it will be fast.
But, in case of memory screen it will add code complexity. For example, I will need to verify that filter value is default. And there are many others like this.

Reply:
We wouldn't need to check the filter value for this case. A good check here would be that chart data is empty, profile data is empty, and diff snapshots is empty. The offline import framework does a check if data is empty and shows an error notification. It's not critical, but would be nice to have for consistency.

Reply:
Hm.
If user did not do anything with screen, but crashed and selected to see historical data they will see error?
No filter? What if user typed something smart into filter and crashed, and just wants to copy/paste this value?

Reply:
Actually, memory screen is never empty, because user always see profile data.